app/browser/api/ledger.js

```diff
@@ -0,0 +1,2187 @@
+  const reconcileStamp = ledgerState.getInfoProp(state, 'reconcileStamp')
+
+  if (!reconcileStamp) {
+    return
+  }
+
+  if (reconcileStamp - new Date().getTime() < miliseconds.day) {
+    if (sufficientBalanceToReconcile(state)) {
+      if (shouldShowNotificationReviewPublishers()) {
+        const reconcileFrequency = ledgerState.getInfoProp(state, 'reconcileFrequency')
+        showNotificationReviewPublishers(reconcileStamp + ((reconcileFrequency - 2) * miliseconds.day))
+      }
+    } else if (shouldShowNotificationAddFunds()) {
+      showNotificationAddFunds()
+    }
+  } else if (reconcileStamp - new Date().getTime() < 2 * miliseconds.day) {
+    if (sufficientBalanceToReconcile(state) && (shouldShowNotificationReviewPublishers())) {
+      showNotificationReviewPublishers(new Date().getTime() + miliseconds.day)
+    }
+  }
+}
+
+const showDisabledNotifications = (state) => {
+  if (!getSetting(settings.PAYMENTS_NOTIFICATION_TRY_PAYMENTS_DISMISSED)) {
+    const firstRunTimestamp = state.get('firstRunTimestamp')
+    if (new Date().getTime() - firstRunTimestamp < appConfig.payments.delayNotificationTryPayments) {
+      return
+    }
+
+    appActions.showNotification({
+      greeting: locale.translation('updateHello'),
+      message: locale.translation('notificationTryPayments'),
+      buttons: [
+        {text: locale.translation('noThanks')},
+        {text: locale.translation('notificationTryPaymentsYes'), className: 'primaryButton'}
+      ],
+      options: {
+        style: 'greetingStyle',
+        persist: false
+      }
+    })
+  }
+}
+
+const showNotifications = (state) => {
+  if (getSetting(settings.PAYMENTS_ENABLED)) {
+    if (getSetting(settings.PAYMENTS_NOTIFICATIONS)) {
+      showEnabledNotifications(state)
+    }
+  } else {
+    showDisabledNotifications(state)
+  }
+}
+
+const cacheRuleSet = (state, ruleset) => {
+  if (!ruleset || underscore.isEqual(_internal.ruleset.raw, ruleset)) {
+    return state
+  }
+
+  try {
+    let stewed = []
+    ruleset.forEach((rule) => {
+      let entry = {condition: acorn.parse(rule.condition)}
+
+      if (rule.dom) {
+        if (rule.dom.publisher) {
+          entry.publisher = {
+            selector: rule.dom.publisher.nodeSelector,
+            consequent: acorn.parse(rule.dom.publisher.consequent)
+          }
+        }
+        if (rule.dom.faviconURL) {
+          entry.faviconURL = {
+            selector: rule.dom.faviconURL.nodeSelector,
+            consequent: acorn.parse(rule.dom.faviconURL.consequent)
+          }
+        }
+      }
+      if (!entry.publisher) entry.consequent = rule.consequent ? acorn.parse(rule.consequent) : rule.consequent
+
+      stewed.push(entry)
+    })
+
+    _internal.ruleset.raw = ruleset
+    _internal.ruleset.cooked = stewed
+    if (!synopsis) {
+      return state
+    }
+
+    let syncP = false
+    const publishers = ledgerState.getPublishers(state)
+    for (let item of publishers) {
```

Comment:
another one which can benefit from index based for loop.  To get a comparison of each performance, you can check out https://jsperf.com/fastest-array-loops-in-javascript/24

Reply:
answered above